src/aes.rs

```diff
@@ -0,0 +1,271 @@
+use super::Error;
+use crate::{
+    create_signed_header,
+    icl_header_v4::{self, v4document_header::edek_wrapper::Edek},
+    EncryptedDocument, EncryptedPayload, EncryptionKey, PlaintextDocument, MAGIC, V0,
+};
+use aes_gcm::{aead::Aead, aead::Payload, AeadCore, Aes256Gcm, KeyInit, Nonce};
+use bytes::Bytes;
+use rand::{CryptoRng, RngCore};
+
+type Result<T> = core::result::Result<T, super::Error>;
+const DETACHED_HEADER_LEN: usize = 5;
+const IV_LEN: usize = 12;
+
+pub fn generate_aes_edek<R: CryptoRng + RngCore>(
+    rng: &mut R,
+    kek: EncryptionKey,
+    id: &str,
+) -> Result<(
+    EncryptionKey,
+    icl_header_v4::v4document_header::edek_wrapper::Aes256GcmEncryptedDek,
+)> {
+    let document_key = {
+        let mut buffer = [0u8; 32];
+        rng.fill_bytes(&mut buffer);
+        buffer
+    };
+    let (iv, edek) = aes_encrypt(kek, &document_key, &[], rng)?;
+    let aes_edek = icl_header_v4::v4document_header::edek_wrapper::Aes256GcmEncryptedDek {
+        ciphertext: edek.0.into(),
+        iv: Bytes::copy_from_slice(&iv),
+        id: id.into(),
+        ..Default::default()
+    };
+    Ok((EncryptionKey(document_key), aes_edek))
+}
+
+/// Generate an aes edek, and encrypt it using the kek. The provided id will be put into the Aes256GcmEdek.
+/// The edek will be placed into a V4DocumentHeader and the signature will be computed.
+/// The generated aes dek is the key used to compute the signature.
+pub fn generate_aes_edek_and_sign<R: CryptoRng + RngCore>(
+    rng: &mut R,
+    kek: EncryptionKey,
+    id: &str,
+) -> Result<(EncryptionKey, icl_header_v4::V4DocumentHeader)> {
+    let (aes_dek, aes_edek) = generate_aes_edek(rng, kek, id)?;
+    Ok((
+        aes_dek,
+        create_signed_header(
+            icl_header_v4::v4document_header::EdekWrapper {
+                edek: Some(Edek::Aes256GcmEdek(aes_edek)),
+                ..Default::default()
+            },
+            aes_dek,
+        ),
+    ))
+}
+
+/// Decrypt the aes edek. Does not verify signature of the header or check that the id is appropriate.
+/// You must do that as a separate step.
+pub fn decrypt_aes_edek(
+    kek: &EncryptionKey,
+    aes_edek: &icl_header_v4::v4document_header::edek_wrapper::Aes256GcmEncryptedDek,
+) -> Result<EncryptionKey> {
+    let iv = aes_edek.iv.as_ref().try_into().map_err(|_| {
+        Error::DecryptError("IV from the edek was not the correct length.".to_string())
+    })?;
+    aes_decrypt(kek, iv, &aes_edek.ciphertext, &[])
+        .and_then(|dek_bytes| {
+            dek_bytes
+                .try_into()
+                .map_err(|_| Error::DecryptError("iv was not of the correct size".to_string()))
```

Comment:
```suggestion
                .map_err(|_| Error::DecryptError("IV was not of the correct size".to_string()))
```

Is this actually possible with the `try_into()` to get the IV just before?

Reply:
Good catch. The error message was wrong. The AES DEK isn't the correct size if this try_into fails.